src/Service/Terminal/Service_Terminal.re

```diff
@@ -199,18 +199,50 @@ module Effect = {
     |> List.to_seq
     |> Hashtbl.of_seq;
 
+  let controlKeyRegex = Oniguruma.OnigRegExp.create("<C-([a-z])>");
+
+  let getControlKey = keyCandidate => {
+    Oniguruma.(
+      controlKeyRegex
+      |> Stdlib.Result.map(OnigRegExp.search(keyCandidate, 0))
+      |> Stdlib.Result.to_option
+      |> Utility.OptionEx.flatMap(matches =>
+           if (Array.length(matches) < 2) {
+             None;
+           } else {
+             let key = OnigRegExp.Match.getText(matches[1]);
+             if (String.length(key) == 1) {
+               Some(key.[0]);
+             } else {
+               None;
+             };
+           }
+         )
+    );
+  };
+
```

Comment:
This seems way more complex than it should be. It'd be nice to have a general key parser, instead of using the vim-style strings as our input currency.

Reply:
I talked about this with @fanantoxa a while ago. Not sure if they're working on it, or wanted to work on it, but I think we concluded that a simplified record-type mostly carrying the same information as the string, should work well. We can then easily serialize it at the vim boundary, while passing it around as a record elsewhere.

Reply:
That sounds great! Something like:
```
type key = {
   controlKey: bool,
   altKey: bool,
   shiftKey: bool,
   metaKey: bool,
   keyCode: int,  // SDL2 keycode
   scanCode: int, // Physical scan code
};
```

This would make all of our input handling much more flexible (ie, we could handle VSCode style bindings like `Command+P`)

Reply:
The problem is that we also generate presses from `textInput` events, which don't include key codes and scan codes. So those would have to be optional, and we might have an additional `key: string` instead.

Reply:
Good point - we'd need our model to accomodate the text case as well.